Experimental/src/Ranges/ImmutableRangeList.cs

```diff
@@ -0,0 +1,273 @@
+        {
+            string result;
+
+            if (IsEmpty)
+            {
+                result = string.Empty;
+            }
+            else if (IsFull)
+            {
+                result = "...";
+            }
+            else
+            {
+                result = string.Join(", ", Values.Select(v => v.DisplayValue()));
+            }
+
+            return "[" + result + "]";
+        }
+
+        public ImmutableRangeList<TValue> Add(Range<TValue> range)
+        {
+            if (range.IsEmpty)
+                return this;
+
+            if (Values == null)
+                return new ImmutableRangeList<TValue>(range);
+
+            return this.Concat(new[] {range}).ToRangeList();
+        }
+
+        public ImmutableRangeList<TValue> AddRange(IEnumerable<Range<TValue>> ranges)
+        {
+            if (ranges == null)
+                return this;
+
+            return Values == null ? ranges.ToRangeList() : Values.Concat(ranges).ToRangeList();
+        }
+
+        public ImmutableRangeList<TValue> RemoveRange(IEnumerable<Range<TValue>> ranges)
+        {
+            if (Values == null)
+                return this;
+
+            if (ranges == null)
+                return this;
+
+            var rangesList = ranges.ToList();
+
+            var sourceList = Values;
+            List<Range<TValue>> copyValues = null;
+
+            foreach (var toRemove in rangesList)
+            {
+                int i = 0;
+                do
+                {
+                    var current = sourceList[i];
+                    if (current.Intersects(toRemove))
+                    {
+                        var removed = current.Exclude(toRemove);
+                        if (copyValues == null)
+                        {
+                            copyValues = Values.ToList();
+                            sourceList = copyValues;
+                        }
+                        sourceList.RemoveAt(i);
+                        var prevCount = sourceList.Count;
+                        sourceList.InsertRange(i, removed);
+                        i += sourceList.Count - prevCount;
+                    }
+                    else
+                    {
+                        ++i;
+                    }
+                } while (i < sourceList.Count);
+            }
+
+            if (copyValues == null)
+                return this;
+
+            return copyValues.ToRangeList();
+        }
+
+        public ImmutableRangeList<TValue> Invert()
+        {
+            if (IsFull)
+                return Empty;
+            if (IsEmpty)
+                return Full;
+
+            return Full.RemoveRange(Values);
+        }
+
+        public ImmutableRangeList<TValue> Intersect(IEnumerable<Range<TValue>> ranges)
+        {
+            if (Values == null || ranges == null)
+                return this;
+
+            var intersected = ranges
+                .SelectMany(r => Values, (r, v) => r.Intersect(v))
```

Comment:
I can't see optimization here. Range.Intersect accepts not IEnumerable<Range> but Range parameter.